src/bootstrap/native.rs

```diff
@@ -321,6 +310,38 @@ impl Step for Llvm {
 
         cfg.build();
 
+        // On macOS, we want the build directory in LLVM to match linux with an
+        // unversioned libLLVM.dylib and versioned
+        // `libLLVM-$MAJOR_VERSION-$VERSION_SUFFIX.dylib`. This is necessary for
+        // llvm-config to work, for example. Unfortunately, LLVM itself does not
+        // currently produce the versioned library name despite llvm-config
+        // expecting it, so for now just manually create it.
+        if target.contains("apple-darwin") && builder.config.llvm_link_shared {
+            let libdir = out_dir.join("lib");
+            let llvm_dylib_path = libdir.join("libLLVM.dylib");
+            if let Some(suffix) = crate::native::llvm_version_suffix(builder) {
+                let mut cmd = Command::new(&build_llvm_config);
```

Comment:
I'm not sure I understand what's going on here. It seems like you're running `llvm-config` successfully here? But then you also need to copy around dylibs to make it run successfully? How does this work out?

Reply:
llvm-config can run with --version but will either not run or report "incorrect" information with other options. It isn't actually "linked" to libLLVM.dylib or the versioned dylib, it just does some directory traversal to find them beside itself at runtime with some commands, erroring out if it can't. If we don't do this we get the error in https://github.com/rust-lang/rust/pull/76810#issuecomment-693795808.